File: 1200/1679B/main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() 
{
    int n, q;
    cin >> n >> q;

    vector<long long> lastVal(n + 1);
    vector<int> lastTime(n + 1, 0);

    long long sum = 0;
    for (int i = 1; i <= n; i++) 
    {
        cin >> lastVal[i];
        sum += lastVal[i];
    }

    long long globalVal = 0;
    int globalTime = 0;

    for (int t = 1; t <= q; t++) 
    {
        int type;
        cin >> type;

        if (type == 1) 
        {
            int idx;
            long long x;
            cin >> idx >> x;

            long long oldVal;
            if (lastTime[idx] >= globalTime)
                oldVal = lastVal[idx];
            else
                oldVal = globalVal;

            sum = sum - oldVal + x;
            lastVal[idx] = x;
            lastTime[idx] = t;
        } 
        else 
        {
            long long x;
            cin >> x;
            globalVal = x;
            globalTime = t;
            sum = 1LL * n * x;
        }

        cout << sum << "\n";
    }

    return 0;
}

File: 1200/1679B/main.java
import java.util.*;
public class main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int q = sc.nextInt();
 
        long[] lastVal = new long[n + 1];
        int[] lastTime = new int[n + 1];
 
        long sum = 0;
        for (int i = 1; i <= n; i++) 
        {
            lastVal[i] = sc.nextLong();
            sum += lastVal[i];
            // lastTime[i] = 0 initially
        }
 
        long globalVal = 0;
        int globalTime = 0;
 
        for (int t = 1; t <= q; t++) 
        {
            int type = sc.nextInt();
            if (type == 1) 
            {
                int idx = sc.nextInt();
                long x = sc.nextLong();
 
                long oldVal;
                // If this element was updated after last global reset
                if (lastTime[idx] >= globalTime)
                    oldVal = lastVal[idx];
                else 
                {
                    // Otherwise its value was effectively globalVal
                    oldVal = globalVal;
                }
 
                sum = sum - oldVal + x;
                lastVal[idx] = x;
                lastTime[idx] = t; // update its timestamp
            } 
            else 
            {
                // type == 2
                long x = sc.nextLong();
                globalVal = x;
                globalTime = t;
                sum = n * x;
            }
 
            System.out.println(sum);
        }
    }
}